### DISCUSSION TITLE
Renovate unnecessarily rebases PRs

### DISCUSSION THREAD

**Discussion by gkalpak** [renovatebot/renovate#8982] [Request Help]
**What Renovate type, platform and version are you using?**

We are using the Renovate App on GitHub.

**Describe the bug**

On the [angular/ngcc-validation](https://github.com/angular/ngcc-validation) repo, in the last few days (at least), we have noticed the following behavior:

Once Renovate merges one of its PRs, it will rebase all its open PRs, although there should be no merge conflicts created by the recent merge. 

For example, you can see that once https://github.com/angular/ngcc-validation/pull/2643 was just merged, Renovate rebased https://github.com/angular/ngcc-validation/pull/2645 (which had already passed CI). By looking at the changed files, it is clear that there should be no merge conflicts between the two PRs.

This behavior is highly undesirable, since (in addition to taking a lot longer to merge PRs) it causes CI to run much more than necessary (increasing CI costs and consuming CI resources). 
Any insight on what might be going on or how to fix it would be appreciated :pray:

**Relevant debug logs**

I am not sure what part of [the logs][1] is relevant. You can find the whole file [here][2].

There were a couple of places that said that the branch was stale and needed rebasing and the lockfile needed updating, but not sure what to make of that: 

```
DEBUG: Branch is stale and needs rebasing(branch="renovate/azure-msal-browser-2.x")
DEBUG: Using reuseExistingBranch: false(branch="renovate/azure-msal-browser-2.x")
```

```
DEBUG: yarn.lock needs updating(branch="renovate/azure-msal-angular-2.x")
DEBUG: updateYarnOffline resolvedPaths(branch="renovate/azure-msal-angular-2.x")
{
  "resolvedPaths": []
}
DEBUG: Updated 1 lock files(branch="renovate/azure-msal-angular-2.x")
{
  "updatedArtifacts": [
    "yarn.lock"
  ]
}
```

**Have you created a minimal reproduction repository?**

- [ ] This is a really small bug, it does not need a reproduction (think small typo)
- [ ] I have provided a minimal reproduction repository
- [x] I don't have time for that, but it happens in a public repository I have linked to
- [ ] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

N/A

[1]: https://app.renovatebot.com/dashboard#github/angular/ngcc-validation/311679650
[2]: https://github.com/renovatebot/renovate/files/6082778/ngcc-validation-311679650.renovate-logs.txt

**Comment by viceice** [comment]
This is by design, renovate will ever only automerge up to date pr, so it needs to rebase. see [rebaseWhen](https://docs.renovatebot.com/configuration-options/#rebasewhen) to change behavior, but this will stop automerge i think.

Also it will only automerge when renovate is latest git author (can be changed when #8893 is done).

**Reply by gkalpak** [reply]
Thx for the answer. I have a couple of follow-up questions:

1. Has this changed recently (because I am fairly confident that this wasn't the case some time ago)?
2. If setting [rebaseWhen][1] to `conflicted` stops automerging, then is there any point in having `automerge` enabled without [prConcurrentLimit][2] set to `1`?

[1]: https://docs.renovatebot.com/configuration-options/#rebasewhen
[2]: https://docs.renovatebot.com/configuration-options/#prconcurrentlimit

**Reply by viceice** [reply]
> Thx for the answer. I have a couple of follow-up questions:
> 
> 1. Has this changed recently (because I am fairly confident that this wasn't the case some time ago)?

Yes, there was a bug, so renovate never rebased non-conflicted automerge pr's. And with branch-protection (must up-to-date) renovate had never automerged. So we needed to change this 

> 2. If setting [rebaseWhen](https://docs.renovatebot.com/configuration-options/#rebasewhen) to `conflicted` stops automerging, then is there any point in having `automerge` enabled without [prConcurrentLimit](https://docs.renovatebot.com/configuration-options/#prconcurrentlimit) set to `1`?

I'm not sure, maybe `rebaseWhen=conflicted` will do what you want. @rarkins Do you know more ?

**Reply by alan-agius4** [reply]
Thanks @viceice for your help, but isn't `rebaseWhen=conflicted`, the default unless PRs require to be up to date with the base branch?

It is also indicated in the PRs status opened by renovate ex: https://github.com/angular/ngcc-validation/pull/2643 

> ♻️  Rebasing: Whenever PR becomes conflicted, or you tick the rebase/retry checkbox.

Or setting it explicitly has a different behaviour? or maybe the rebasing status is not correct in the PR?

**Reply by viceice** [reply]
default is `rebaseWhen=auto` which will rebase all automerge pr's / branches by default. All other are only rebased on conflict.

This is the recent change we made. so renovate is able to rebase automerge pr, so it's not stopped by up-to-date branch protection

**Reply by viceice** [reply]
maybe we should add this automerge info's to https://docs.renovatebot.com/configuration-options/#rebasewhen @HonkingGoose  @rarkins what do you think?

**Reply by rarkins** [reply]
We should make sure that the description in the PR matches as well, to explain that stale rebasing is necessary if automerge is enabled.

@alan-agius4 would you *want* Renovate to automerge a branch which is not up to date? We have assumed that's a very bad idea and not allowed it to be configurable, but perhaps we shouldn't be so opinionated if there's demand. Right now automerge=true implicitly turns on rebaseWhen=behind-base-branch, which can in theory result in a broken master branch if two Renovate PRs are in progress at the same time but aren't compatible with each other.

I won't really understand question (2) from @gkalpak. rebaseWhen=conflicted doesn't stop automerging (I think), instead automerging overrides rebaseWhen. And while I'd recommend keeping a small concurrent limit of PRs if automerging (because if you have N open and automerging then you need N-1 rebased each time), if you make it too small and one of them is blocked from automerging then you are completely blocked (no new PRs coming).

**Reply by gkalpak** [reply]
Regarding my second question:

So, do I understand correctly that our options currently are:
1. Limit the concurrent branches/PRs (with `prConcurrentLimit`) to 1, which has the downside that one failing PR blocks all updates.
2. Keep concurrent branch/PR limit above 1, which has the downside that we will run CI `(n * (n + 1)) / 2` times for `n` concurrent PRs.

**Reply by rarkins** [reply]
Consider using `dependencyDashboard`. It would let you:
- Still have visibility of an update if you close it because for some reason it's blocked and you can't work on it right now,  and
- Override the concurrent limit by ticking the checkbox for any rate-limited PRs to have the PR created on-demand

I also see there's some potential for additional grouping (which reduces concurrent PRs).

It's still on the table to support functionality to allow automerging from a non-up-to-date branch, if you think that's a good idea.

Finally, consider GitHub Actions - currently unlimited minutes for public repos :)

**Reply by gkalpak** [reply]
Using the `dependencyDashboard` won't really solve the issue for us for a couple of reasons:
1. We want the process to be as automated as possible.
2. Manually triggering more concurrent PRs won't buy us much if each PR has to be rebased after another PR is merged.

In any case, thx for the info :+1:
I wanted to make sure I understand our options. We'll investigate what route makes the most sense for `angular/ngcc-validation`. Maybe we'll start with limiting concurrency and see how it goes.

(BTW, having an option to auto-merge non-up-to-date branches would work for `angular/ngcc-validation`, but that argueably not a typical usecase. So not sure if it is worth adding to Renovate.)

**Reply by rarkins** [reply]
We aim to have great defaults, but otherwise not be overly opinionated about what users "should" do. So we should think about how we can enable it without contributing to any other users from shooting themselves in the foot. I think if the following holds:
- rebaseWhen=conflicted
- automerge=true
- No GitHub branch protection enforcing up-to-date branches

Then we can respect the rebaseWhen=conflicted instead of converting it to rebaseWhen=behind-base-branch

**Reply by alan-agius4** [reply]
> @alan-agius4 would you want Renovate to automerge a branch which is not up to date? We have assumed that's a very bad idea and not allowed it to be configurable, but perhaps we shouldn't be so opinionated if there's demand. Right now automerge=true implicitly turns on rebaseWhen=behind-base-branch, which can in theory result in a broken master branch if two Renovate PRs are in progress at the same time but aren't compatible with each other.

I do understand why this is a bad idea, but in some cases, with this approach it might end being a limiting factor when there a lot are of PRs from renovate and CI takes a considerable amount of time. I was thinking of proposing to enable automerging in the Angular CLI repo, but understanding how this works it off the table. As rebasing every PR, when another one gets merged will be to wasteful and time consuming (Our entire CI takes approx 30mins).

I think what you mentioned above to respect `rebaseWhen=conflicted` makes total sense. Also, I think that this behaviour should be stated in the docs and maybe even in the PR status, as it's not clear that `rebaseWhen` is not `conflicted` when using automerge.

**Reply by toxsick** [reply]
Hey guys,

good I found this. If this is new behaviour the docs on this are a little misleading I think:

https://docs.renovatebot.com/updating-rebasing/#rebasing-out-of-date-branches

I would think that renovate will not by default rebase every merge request, when something else is committed to the master branch. In my case I definitely have to disable this setting and use the `"rebaseWhen": "conflicted"`, because I have multiple CI Pipelines with e2e tests end everything, which run up to 5 times before they get finally merged, because they play ping pong with other merge requests.

But now I know about the new default I'm happy :)

Thanks a lot for renovate guys, saves me hours every week!

**Reply by HonkingGoose** [reply]
You might want to follow issue #8988 to be notified of changes to the docs. 😉

**Reply by northern-mindset** [reply]
Interestingly, #8988 mentions “update text in Renovate PR body” in the title, but the “Update Renovate PR body text” checkbox not ticked in the description of the issue. Perhaps because of this @rarkins have just updated the docs, as questioned by @HonkingGoose in https://github.com/renovatebot/renovate/issues/8988#issuecomment-792742940.

Also, #9031 effectively only adds a warning to the docs that,

> `rebaseWhen=conflicted` is not recommended if you have enabled Renovate automerge

It does not add anything regarding the fact that “`rebaseWhen=auto` will rebase all automerge pr's / branches by default.”

However, https://docs.renovatebot.com/updating-rebasing/ describes the behavior correctly, sadly I could not find that page and spent too much time trying to figure out why Renovate is rebasing my branches.

I think that changing the PR text is absolutely necessary.

**Reply by northern-mindset** [reply]
> Then we can respect the rebaseWhen=conflicted instead of converting it to rebaseWhen=behind-base-branch

@rarkins is this a plan or reality now?

**Reply by rarkins** [reply]
It's done, but it's still a bad idea which will block automerging in most cases.
